File: lsd-interceptors/src/test/java/io/lsdconsulting/interceptors/http/common/DefaultHttpInteractionHandlerTest.java
package io.lsdconsulting.interceptors.http.common;

import com.lsd.core.LsdContext;
import com.lsd.core.domain.Message;
import com.lsd.core.domain.MessageType;
import com.lsd.core.domain.Participant;
import com.lsd.core.domain.SequenceEvent;
import io.lsdconsulting.interceptors.common.HeaderKeys;
import io.lsdconsulting.interceptors.http.naming.DestinationNameMappings;
import io.lsdconsulting.interceptors.http.naming.SourceNameMappings;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.time.Duration;
import java.util.Map;

import static com.lsd.core.domain.ParticipantType.PARTICIPANT;
import static java.util.Collections.emptyMap;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

class DefaultHttpInteractionHandlerTest {

    private final Map<String, String> serviceNameHeaders = Map.of(
            HeaderKeys.TARGET_NAME.key(), "bob",
            HeaderKeys.SOURCE_NAME.key(), "juliet"
    );
    private final SourceNameMappings sourceNameMapping = path -> "andrea";
    private final DestinationNameMappings destinationNameMapping = path -> "bren";

    private final ArgumentCaptor<SequenceEvent> messageCaptor = ArgumentCaptor.forClass(SequenceEvent.class);

    private final LsdContext lsdContext = Mockito.spy(LsdContext.class);

    private final DefaultHttpInteractionHandler handler = new DefaultHttpInteractionHandler(lsdContext, sourceNameMapping, destinationNameMapping);
    private final Participant bob = PARTICIPANT.called("bob");
    private final Participant juliet = PARTICIPANT.called("juliet");
    private final Participant bren = PARTICIPANT.called("bren");
    private final Participant andrea = PARTICIPANT.called("andrea");

    @Test
    void usesTestStateToLogRequest() {
        handler.handleRequest("GET", emptyMap(), "/path", "{\"type\":\"request\"}");

        verify(lsdContext, times(1)).capture(messageCaptor.capture());
        var message = extractFirstMessageFromCaptor();

        assertThat(message.getFrom()).isEqualTo(andrea);
        assertThat(message.getTo()).isEqualTo(bren);
        assertThat(message.getLabel()).isEqualTo("GET /path");
        assertThat(message.getType()).isEqualTo(MessageType.SYNCHRONOUS);
        assertThat(message.getData().toString())
                .contains("<p>{\n  &quot;type&quot;: &quot;request&quot;\n}</p>");
    }

    @Test
    void usesTestStateToLogResponse() {
        handler.handleResponse("200 OK", emptyMap(), emptyMap(), "/path", "response body", Duration.ofMillis(5));

        verify(lsdContext, times(1)).capture(messageCaptor.capture());
        var message = extractFirstMessageFromCaptor();

        assertThat(message.getFrom()).isEqualTo(bren);
        assertThat(message.getTo()).isEqualTo(andrea);
        assertThat(message.getLabel()).isEqualTo("200 OK (5ms)");
        assertThat(message.getType()).isEqualTo(MessageType.SYNCHRONOUS_RESPONSE);
        assertThat(message.getData().toString()).contains("<p>response body</p");
        assertThat(message.getDuration()).isEqualTo(Duration.ofMillis(5));
    }

    @Test
    void headerValuesForSourceAndDestinationArePreferredWhenLoggingRequest() {
        handler.handleRequest("GET", serviceNameHeaders, "/path", "");

        verify(lsdContext, times(1)).capture(messageCaptor.capture());
        var message = extractFirstMessageFromCaptor();

        assertThat(message.getFrom()).isEqualTo(juliet);
        assertThat(message.getTo()).isEqualTo(bob);
        assertThat(message.getLabel()).isEqualTo("GET /path");
        assertThat(message.getType()).isEqualTo(MessageType.SYNCHRONOUS);
        assertThat(message.getData().toString())
                .contains("Source-Name: juliet")
                .contains("Target-Name: bob");
    }

    @NotNull
    private Message extractFirstMessageFromCaptor() {
        return messageCaptor.getAllValues().stream()
                .filter(Message.class::isInstance)
                .map(Message.class::cast)
                .findFirst().orElseThrow();
    }

    @Test
    void headerValuesForSourceAndDestinationArePreferredWhenLoggingResponse() {
        handler.handleResponse("200 OK", serviceNameHeaders, emptyMap(), "/path", "response body", Duration.ofMillis(3));

        verify(lsdContext, times(1)).capture(messageCaptor.capture());
        var message = extractFirstMessageFromCaptor();

        assertThat(message.getFrom()).isEqualTo(bob);
        assertThat(message.getTo()).isEqualTo(juliet);
        assertThat(message.getLabel()).isEqualTo("200 OK (3ms)");
        assertThat(message.getType()).isEqualTo(MessageType.SYNCHRONOUS_RESPONSE);
        assertThat(message.getDuration()).isEqualTo(Duration.ofMillis(3));
        assertThat(message.getData().toString())
                .contains("<h3>Request Headers</h3>")
                .contains("Target-Name: bob")
                .contains("Source-Name: juliet")
                .contains("<p>response body</p>");
    }
}

File: lsd-interceptors/src/test/kotlin/io/lsdconsulting/interceptors/http/common/DefaultHttpInteractionHandlerTest.kt
package io.lsdconsulting.interceptors.http.common

import com.lsd.core.LsdContext
import com.lsd.core.domain.Message
import com.lsd.core.domain.MessageType
import com.lsd.core.domain.ParticipantType
import com.lsd.core.domain.SequenceEvent
import io.lsdconsulting.interceptors.common.HeaderKeys
import io.lsdconsulting.interceptors.http.naming.DestinationNameMappings
import io.lsdconsulting.interceptors.http.naming.SourceNameMappings
import io.mockk.every
import io.mockk.slot
import io.mockk.spyk
import io.mockk.verify
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.Duration

internal class DefaultHttpInteractionHandlerTest {
    private val serviceNameHeaders = mapOf(
        HeaderKeys.TARGET_NAME.key() to "bob",
        HeaderKeys.SOURCE_NAME.key() to "juliet",
    )
    private val sourceNameMapping = SourceNameMappings { "andrea" }
    private val destinationNameMapping = DestinationNameMappings { "bren" }
    private val messageSlot = slot<SequenceEvent>()
    private val lsdContext = spyk<LsdContext>()
    private val handler = DefaultHttpInteractionHandler(lsdContext, sourceNameMapping, destinationNameMapping)
    private val bob = ParticipantType.PARTICIPANT.called("bob")
    private val juliet = ParticipantType.PARTICIPANT.called("juliet")
    private val bren = ParticipantType.PARTICIPANT.called("bren")
    private val andrea = ParticipantType.PARTICIPANT.called("andrea")

    @BeforeEach
    fun setup() {
        every { lsdContext.capture(capture(messageSlot), any()) } returns Unit
    }

    @Test
    fun usesTestStateToLogRequest() {
        handler.handleRequest("GET", emptyMap(), "/path", "{\"type\":\"request\"}")

        verify { lsdContext.capture(any(), any()) }
        val (_, from, to, label, type, _, data) = extractFirstMessageFromCaptor()

        assertThat(from).isEqualTo(andrea)
        assertThat(to).isEqualTo(bren)
        assertThat(label).isEqualTo("GET /path")
        assertThat(type).isEqualTo(MessageType.SYNCHRONOUS)
        assertThat(data.toString())
            .contains("<p>{\n  &quot;type&quot;: &quot;request&quot;\n}</p>")
    }

    @Test
    fun usesTestStateToLogResponse() {
        handler.handleResponse("200 OK", emptyMap(), emptyMap(), "/path", "response body", Duration.ofMillis(5))

        verify { lsdContext.capture(any(), any()) }
        val (_, from, to, label, type, _, data, duration) = extractFirstMessageFromCaptor()

        assertThat(from).isEqualTo(bren)
        assertThat(to).isEqualTo(andrea)
        assertThat(label).isEqualTo("200 OK (5ms)")
        assertThat(type).isEqualTo(MessageType.SYNCHRONOUS_RESPONSE)
        assertThat(data.toString()).contains("<p>response body</p")
        assertThat(duration).isEqualTo(Duration.ofMillis(5))
    }

    @Test
    fun headerValuesForSourceAndDestinationArePreferredWhenLoggingRequest() {
        handler.handleRequest("GET", serviceNameHeaders, "/path", "")

        verify { lsdContext.capture(any(), any()) }
        val (_, from, to, label, type, _, data) = extractFirstMessageFromCaptor()

        assertThat(from).isEqualTo(juliet)
        assertThat(to).isEqualTo(bob)
        assertThat(label).isEqualTo("GET /path")
        assertThat(type).isEqualTo(MessageType.SYNCHRONOUS)
        assertThat(data.toString())
            .contains("Source-Name: juliet")
            .contains("Target-Name: bob")
    }

    @Test
    fun headerValuesForSourceAndDestinationArePreferredWhenLoggingResponse() {
        handler.handleResponse("200 OK", serviceNameHeaders, emptyMap(), "/path", "response body", Duration.ofMillis(3))

        verify { lsdContext.capture(any(), any()) }
        val (_, from, to, label, type, _, data, duration) = extractFirstMessageFromCaptor()

        assertThat(from).isEqualTo(bob)
        assertThat(to).isEqualTo(juliet)
        assertThat(label).isEqualTo("200 OK (3ms)")
        assertThat(type).isEqualTo(MessageType.SYNCHRONOUS_RESPONSE)
        assertThat(duration).isEqualTo(Duration.ofMillis(3))
        assertThat(data.toString())
            .contains("<h3>Request Headers</h3>")
            .contains("Target-Name: bob")
            .contains("Source-Name: juliet")
            .contains("<p>response body</p>")
    }

    private fun extractFirstMessageFromCaptor(): Message =
        messageSlot.captured
            .takeIf { obj: SequenceEvent -> Message::class.java.isInstance(obj) }
            ?.let { obj: SequenceEvent -> Message::class.java.cast(obj) }
            ?: throw Exception()
}
